crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopInvariant.hs

```diff
@@ -0,0 +1,1151 @@
+simpleLoopInvariant sym loopNum cfg@C.CFG{..} mem_var loop_invariant = do
+  let ?ptrWidth = knownNat @64
+
+  --verbSetting <- W4.getOptionSetting W4.verbosity $ W4.getConfiguration sym
+  --verb <- fromInteger <$> W4.getOpt verbSetting
+
+  loop_map <- computeLoopMap loopNum (C.cfgWeakTopologicalOrdering cfg)
+
+  fixpoint_state_ref <- newIORef @(FixpointState p sym ext rtp blocks) BeforeFixpoint
+
+  putStrLn "Setting up simple loop fixpoints feature."
+  putStrLn ("Loop map: " ++ show loop_map)
+  putStrLn ("WTO: " ++ show (C.cfgWeakTopologicalOrdering cfg))
+  putStrLn (show (C.cfgHandle cfg))
+  writeFile "bigcfg.txt" (show (C.ppCFG' True (C.postdomInfo cfg) cfg))
+
+  return $ C.ExecutionFeature $ \exec_state -> do
+    let ?logMessage = \msg -> do -- when (verb >= (3 :: Natural)) $ do
+          let h = C.printHandle $ C.execStateContext exec_state
+          System.IO.hPutStrLn h msg
+          System.IO.hFlush h
+
+    fixpoint_state <- readIORef fixpoint_state_ref
+    C.withBackend (C.execStateContext exec_state) $ \bak ->
+     case exec_state of
+      C.RunningState (C.RunBlockStart block_id) sim_state
+        | C.SomeHandle cfgHandle == C.frameHandle (sim_state ^. C.stateCrucibleFrame)
+
+        -- make sure the types match
+        , Just Refl <- W4.testEquality
+            (fmapFC C.blockInputs cfgBlockMap)
+            (fmapFC C.blockInputs $ C.frameBlockMap $ sim_state ^. C.stateCrucibleFrame)
+
+          -- loop map is what we computed above, is this state at a loop header
+        , Map.member (C.Some block_id) loop_map ->
+
+            advanceFixpointState bak mem_var loop_invariant block_id sim_state fixpoint_state fixpoint_state_ref
+
+        | otherwise -> do
+            ?logMessage $ "SimpleLoopInvariant: RunningState: RunBlockStart: " ++ show block_id ++ " " ++ show (C.frameHandle (sim_state ^. C.stateCrucibleFrame))
+            return C.ExecutionFeatureNoChange
+
+      -- TODO: maybe need to rework this, so that we are sure to capture even concrete exits from the loop.
+      C.SymbolicBranchState branch_condition true_frame false_frame _target sim_state
+          | Just FixpointRecord{ fixpointBlockId = fixpoint_block_id } <- fixpointRecord fixpoint_state
+          , Just loop_body_some_block_ids <- Map.lookup (C.Some fixpoint_block_id) loop_map
+          , JustPausedFrameTgtId true_frame_some_block_id <- pausedFrameTgtId true_frame
+          , JustPausedFrameTgtId false_frame_some_block_id <- pausedFrameTgtId false_frame
+          , C.SomeHandle cfgHandle == C.frameHandle (sim_state ^. C.stateCrucibleFrame)
+          , Just Refl <- W4.testEquality
+              (fmapFC C.blockInputs cfgBlockMap)
+              (fmapFC C.blockInputs $ C.frameBlockMap $ sim_state ^. C.stateCrucibleFrame)
+          , elem true_frame_some_block_id loop_body_some_block_ids /= elem false_frame_some_block_id loop_body_some_block_ids -> do
+
+            (loop_condition, inside_loop_frame, _outside_loop_frame) <-
```

Comment:
Is this just normalizing the order of the conditions? If the outside loop frame is never used, should this just be deleted in this binding?

Reply:
We can probably delete the unused outside loop frame.  Previously, the code was using it and that got refactored away, but I never pushed it into here.